<div class="alert alert-info"><big><b>Комментарий ревьюера</b></big><br>Привет! 👋<br><br> Меня зовут Дмитрий Кирсанов. <br><br>Комментарии всё ещё обозначены пометкой <b>Комментарий ревьюера</b> и всё ещё содержат описание <br><span class="alert-success">✔️ хорошего, 👍 отличного, 📝 советов</span>,&nbsp;&nbsp;<span class="alert-warning">📌 важного, ⚠️ замечаний, 💡 рекомендаций</span>,&nbsp;&nbsp;<span class="alert-danger">🔧 доработок, 🚫 ошибок или 👀 вопросов.</span><br><br>После того, как замечание будет устранено, я сменю красный цвет на жёлтый, чтобы тебе было удобнее выделять те места, которые требуют доработки. Код, которым я предлагаю выделять твои комментарии<!--
НЕ НАДО КОПИРОВАТЬ КОД, ПРИ ПОМОЩИ КОТОРОГО Я ВЫВОЖУ КОД-->:<br><code>&lt;div class=&quot;alert btn-primary&quot;&gt;&lt;b&gt;Комментарий студента:&lt;/b&gt;&lt;br&gt;   Ок   &lt;/div&gt;</code><br><!--
КОПИРОВАТЬ НАДО, КОГДА ВИДЕН ОФОРМЛЕННЫЙ ТЕКСТ-->Всё ещё просьба задавать вопросы и описывать изменения в проект.</div>

<div class="alert alert-info"> <b>Комментарии студента:</b><br>
    Привет, Дмитрий!<br><br>
    Рад тебя видеть!<br>
</div>

### Опишем цели исследования:

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Моя первая задача как аналитика — проанализировать базу данных.

В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных
#### Таблица books:
Содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

#### Таблица authors:
Содержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.

#### Таблица publishers:
Содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства.

#### Таблица ratings:
Содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

#### Таблица reviews:
Содержит данные о пользовательских обзорах:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

### Подключение к базе данных

In [ ]:
# !pip install sqlalchemy

<div class="alert alert-warning"><b>Комментарий ревьюера</b>&nbsp;🔧<br>
Я заметил что первая ячейка имела номер выполнения, отличный от <tt>[1]</tt>. Это может приводить к тому, что не вызовут ошибку какие-то изменения кода, которые вызвали бы ошибку после перезапуска ноутбука (или при его запуске на другой машине), например потому, что в памяти сохраняются переменные, которые были сформированы до того, как код был изменён. Поэтому важно выполнять ячейки с первой по последнюю строго последовательным образом перед тем, как отправлять их кому-то.
</div>

<div class="alert alert-info"> <b>Комментарии студента:</b><br>
    Прощу прощения, перед отправкой буду перезапускать в обязательном порядке<br>
</div>

In [ ]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [ ]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],db_config['pwd'],db_config['host'],db_config['port'],db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<div class="alert alert-success"><b>Комментарий ревьюера</b>&nbsp;📝<br>
Это код из справки, но его можно улучшить. Если задать имена ключам в текстовом шаблоне, то можно гораздо короче и удобнее формировать строки:

    'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
</div>

<div class="alert alert-info"> <b>Комментарии студента:</b><br>
    Спасибо! Пометил себе.
</div>

### Исследуем таблицы — выведем первые строки:

In [ ]:
# выведем первые пять строк таблицы books - она содержит данные о книгах:
query = '''
SELECT *
FROM books
LIMIT 5 '''

In [ ]:
engine.execute(query);

In [ ]:
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<div class="alert alert-success"><b>Комментарий ревьюера</b>&nbsp;<br>
✔️ Хорошо, что ограничиваешь число загружаемых в память локальной машины строк средствами SQL, для крупных таблиц это будет критически важно

Круто что ты ниже отдельным запросом смотришь на другие статистики
</div>

In [ ]:
# выведем наибольшее и наименьшее кол-во страниц и самую раннюю и самую позднюю даты публикации, кол-во строк:
query = '''
SELECT  MAX(num_pages) AS max_num_pages,
        MIN(num_pages) AS min_num_pages,
        MAX(publication_date) AS max_date,
        MIN(publication_date) AS min_date,
        COUNT(*)
FROM books
'''

In [ ]:
engine.execute(query);

In [ ]:
pd.io.sql.read_sql(query, con = engine)

max_num_pages  min_num_pages    max_date    min_date  count
0           2690             14  2020-03-31  1952-12-01   1000

In [ ]:
# далее выведем пять первых строк таблицы authors - она содержит данные об авторах:
query = '''
SELECT *
FROM authors
LIMIT 5 '''

In [ ]:
engine.execute(query);

In [ ]:
pd.io.sql.read_sql(query, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [ ]:
# далее выведем пять первых строк таблицы publishers - она содержит данные об издательствах:
query = '''
SELECT *
FROM publishers
LIMIT 5 '''

In [ ]:
engine.execute(query);

In [ ]:
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [ ]:
# далее выведем пять первых строк таблицы ratings - она содержит данные о пользовательских оценках книг:
query = '''
SELECT *
FROM ratings
LIMIT 5 '''

In [ ]:
engine.execute(query);

In [ ]:
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [ ]:
# посмотрим максимальную и минимальную оценку книг:
query = '''
SELECT MAX(rating) AS max_rating,
       MIN(rating) AS min_rating
FROM ratings
'''

In [ ]:
engine.execute(query);

In [ ]:
pd.io.sql.read_sql(query, con = engine)

max_rating  min_rating
0           5           1

In [ ]:
# на последок выведем пять первых строк таблицы reviews - она содержит данные о пользовательских обзорах:
query = '''
SELECT *
FROM reviews
LIMIT 5 '''

In [ ]:
engine.execute(query);

In [ ]:
pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<div class="alert alert-success"><b>Комментарий ревьюера</b>&nbsp;📝<br>
Хорошо! Мы познакомились с данными: посмотрели содержимое таблиц и заодно проверили соединение с сервером. Ниже некоторые полезные запросы, которые могут помочь в будущем
</div>

In [ ]:
# КОД РЕВЬЮЕРА
# какие ещё таблицы доступны в базе
display(pd.io.sql.read_sql('''

SELECT *
FROM pg_catalog.pg_tables
WHERE schemaname != 'pg_catalog' AND
      schemaname != 'information_schema';

''', con = engine))


# смотрим на типы столбцов в интересующих таблицах
display(pd.io.sql.read_sql('''
SELECT
    table_name,
    column_name,
    data_type,
    is_nullable
FROM INFORMATION_SCHEMA.COLUMNS
WHERE table_name IN ('books', 'authors', 'publishers', 'ratings', 'reviews');
''', con = engine))

schemaname           tablename         tableowner tablespace  hasindexes  \
0      public              orders    praktikum_admin       None        True   
1      public              visits    praktikum_admin       None        True   
2      public  advertisment_costs    praktikum_admin       None        True   
3      public             authors    praktikum_admin       None        True   
4      public              second  praktikum_student       None       False   
5      public            second_b  praktikum_student       None       False   
6      public          publishers    praktikum_admin       None        True   
7      public              author  praktikum_student       None        True   
8      public             reviews    praktikum_admin       None        True   
9      public             ratings    praktikum_admin       None        True   
10     public               books    praktikum_admin       None        True   

    hasrules  hastriggers  rowsecurity  
0      False        False        False  
1      False        False        False  
2      False        False        False  
3      False         True        False  
4      False        False        False  
5      False        False        False  
6      False         True        False  
7      False        False        False  
8      False         True        False  
9      False         True        False  
10     False         True        False

table_name       column_name data_type is_nullable
0      authors         author_id   integer          NO
1      authors            author      text         YES
2   publishers      publisher_id   integer          NO
3   publishers         publisher      text         YES
4      reviews         review_id   integer          NO
5      reviews           book_id   integer         YES
6      reviews          username      text         YES
7      reviews              text      text         YES
8      ratings         rating_id   integer          NO
9      ratings           book_id   integer         YES
10     ratings          username      text         YES
11     ratings            rating   integer         YES
12       books           book_id   integer          NO
13       books         author_id   integer         YES
14       books             title      text         YES
15       books         num_pages   integer         YES
16       books  publication_date      date         YES
17       books      publisher_id   integer         YES

<div class="alert alert-info"> <b>Комментарии студента:</b><br>
    Опять спасибо! Каждый твой совет делает жизнь аналитика понятнее)
</div>

### Посчитаем, сколько книг вышло после 1 января 2000 года:

In [ ]:
# Сделаем одним SQL-запрос для решения задания, выведем результат запроса в тетрадке:

In [ ]:
display(pd.io.sql.read_sql('''
SELECT COUNT(*)
FROM books
WHERE publication_date > '2000-01-01'
''', con = engine))

count
0    819

In [ ]:
display(pd.io.sql.read_sql('''
SELECT COUNT(*)
FROM books
WHERE publication_date >= '2000-01-01'
''', con = engine))

count
0    821

#### Опишем выводы:
С начала 2000 года вышло 821 книга из присутствующих БД
После 01 января 2000 года вышло 819 книг из присутствующих БД

<div class="alert alert-success"><b>Комментарий ревьюера</b>&nbsp;<br>
✔️ Тут всё верно. Сработал бы и запрос с таким условием:

    publication_date >= '2000-01-01'

Если следовать вопросу буквально, то корректно делать строгое сравнение, не включая книги, опубликованные 1 января 2000 года, но по опыту обычно нужны данные с самого начала года. В подобных ситуациях стоит уточнять что именно нужно или сделать оба варианта, если это невозможно.
    
💡 Желательно сопровождать запрос комментариями (их можно начинать с <tt>--</tt>) о том, что в нём происходит, это хорошая привычка, она очень облегчает чтение запросов, особенно длинных.
</div>

<div class="alert alert-info"> <b>Комментарии студента:</b><br>
    Сделал оба варианта.
</div>

### Для каждой книги посчитайте количество обзоров и среднюю оценку:

In [ ]:
# Тут у тимлида надо спросить: "Какие характеристики книг выводить и сколько книг?"
# Сделаем одним SQL-запрос для решения задания, выведем результат запроса в тетрадке:

In [ ]:
display(pd.io.sql.read_sql('''
WITH

-- посчитаем количество обзоров для каждой книги
count_texts AS (SELECT book_id, COUNT(text) AS count_text
                FROM reviews
                GROUP BY book_id
                ),

-- посчитаем среднюю оценку для каждой книги
avg_ratings AS (SELECT book_id, ROUND(AVG(rating), 2) AS avg_rating
                FROM ratings
                GROUP BY book_id
                )

SELECT b.book_id, author_id, title, num_pages, publication_date, publisher_id, count_text, avg_rating
FROM books AS b
LEFT OUTER JOIN count_texts AS ct ON ct.book_id = b.book_id
LEFT OUTER JOIN avg_ratings AS ar ON ar.book_id = b.book_id
ORDER BY avg_rating DESC -- отсортируем по средней оценке

''', con = engine))

book_id  author_id                                              title  \
0         17         94                      A Dirty Job (Grim Reaper  #1)   
1        610        250                           Tai-Pan (Asian Saga  #2)   
2        553        257            School's Out—Forever (Maximum Ride  #2)   
3        513        178               Piercing the Darkness (Darkness  #2)   
4        902        558  The War of Art: Break Through the Blocks & Win...   
..       ...        ...                                                ...   
995      915        580  The World Is Flat: A Brief History of the Twen...   
996      316        321                  His Excellency: George Washington   
997      371        625                                              Junky   
998      202         91                                      Drowning Ruth   
999      303        291                               Harvesting the Heart   

     num_pages publication_date  publisher_id  count_text  avg_rating  
0          387       2006-03-21           331         4.0        5.00  
1          734       2009-09-01            80         2.0        5.00  
2          409       2006-05-23           176         3.0        5.00  
3          448       2003-06-26            67         2.0        5.00  
4          168       2003-04-01           326         2.0        5.00  
..         ...              ...           ...         ...         ...  
995        616       2006-04-18           106         3.0        2.25  
996        320       2005-11-08           309         2.0        2.00  
997        208       2003-04-01           210         2.0        2.00  
998        368       2001-07-31            33         3.0        2.00  
999        453       1995-04-01           212         2.0        1.50  

[1000 rows x 8 columns]

#### Опишем выводы:
Вывели каждой книги количество обзоров и среднюю оценку, упорядочив по средней оценке

<div class="alert alert-warning"><b>Комментарий ревьюера</b>&nbsp;🔧<br>
Ты обратил внимание на количество выведенных записей и на общее количество книг. Выведено меньше книг, чем есть в базе, а нам нужно вывести все книги. Внимательнее посмотри на то, как ты объединяешь таблицы.
</div><div class="alert alert-warning">💡 Лучше сортировать подобные таблицы по каким-то показателям.</div>

<div class="alert alert-info"> <b>Комментарии студента:</b><br>
    Изменил объединение, спасибо.
</div>

<div class="alert alert-success"><b>Комментарий ревьюера 2&nbsp;<big>👍</big></b></div>

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры:

In [ ]:
# Сделаем однин SQL-запрос для решения задания, выведем результат запроса в тетрадке:

In [ ]:
display(pd.io.sql.read_sql('''
SELECT p.publisher, COUNT (book_id) AS count_books
FROM books AS b
LEFT OUTER JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE num_pages > 50 -- книги в которых больше 50 страниц
GROUP BY p.publisher -- сгруппируем по издателю
ORDER BY count_books DESC -- упорядочим по убыванию по кол-ву книг
LIMIT 1 -- возьмем издателя у которого больше всего выпущено книг
''', con = engine))

publisher  count_books
0  Penguin Books           42

#### Опишем выводы:
Наибольшее число книг толще 50 страниц выпустило издательство Penguin Books, кол-во выпущенных этим издательством книг попадающих под заданный критерий составляет 42.

<div class="alert alert-success"><b>Комментарий ревьюера</b>&nbsp;👍<br>
Совершенно верно, это правильное издательство и оно одно, как нас просили.
</div>

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками:

In [ ]:
# Сделаем однин SQL-запрос для решения задания, выведем результат запроса в тетрадке:

In [ ]:
display(pd.io.sql.read_sql('''
WITH

-- отобрали книги у которых 50 и более оценок и посчитали среднюю оценку таких книг
avg_ratings AS (SELECT MIN(b.author_id) AS author_id,  AVG(rating) AS avg_rating
                FROM ratings AS r
                LEFT OUTER JOIN books AS b ON r.book_id = b.book_id
                GROUP BY b.book_id
                HAVING COUNT(rating_id) >= 50 -- взяли книги у которых более 50 оценок
                )

SELECT author, ROUND(AVG(avg_rating),2) AS avg_avg_rating
FROM avg_ratings AS ar
LEFT OUTER JOIN authors AS a ON a.author_id = ar.author_id  -- ко всем author_id определенным во врем.таблице добавим имя автора
GROUP BY a.author_id  -- сгруппируем по автору
ORDER BY avg_avg_rating DESC --упорядочим по среднему средней оценки по убыванию
LIMIT 1 -- возьмем автора с наивысшей оценкой


''', con = engine))

author  avg_avg_rating
0  J.K. Rowling/Mary GrandPré            4.28

<div class="alert alert-danger"><b>Комментарий ревьюера</b>&nbsp;🔧<br>
Не выведен сам рейтинг и нету группировки по авторам - по сути выведен автор самой рейтинговой книги из тех, что с 50 страницами и более</div>

<div class="alert alert-info"> <b>Комментарии студента:</b><br>
    Исправил, не так прочитал и понял задание. Спасибо.
</div>

<div class="alert alert-success"><b>Комментарий ревьюера 2&nbsp;<big>👍</big></b></div>

#### Опишем выводы:
Автор с самой высокой средней оценкой книг (учитывая только книги с 50 и более оценками) - J.K. Rowling/Mary GrandPré, среднее средних оценок его книг (с 50 и более оценками) - 4.28.

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [ ]:
# Сделаем однин SQL-запрос для решения задания, выведем результат запроса в тетрадке:

In [ ]:
display(pd.io.sql.read_sql('''
SELECT ROUND(AVG(count_text), 2)
FROM (SELECT COUNT(text) AS count_text
      FROM reviews
      WHERE username IN (SELECT username
                         FROM ratings
                         GROUP BY username
                         HAVING COUNT(rating) > 50
                        )
      GROUP BY username
     ) AS count_texts
''', con = engine))

round
0  24.33

#### Опишем выводы:
Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24,33

<div class="alert alert-success"><b>Комментарий ревьюера</b>&nbsp;✔️<br>
И это... правильный ответ!
</div>

<div class="alert alert-warning"><b>Комментарий ревьюера</b>&nbsp;💡<br>
Сам себя не похвалишь, не подчеркнёшь важность и сложность своей работы — никто этого делать за тебя не будет. Выводы — как раз такое место, где можно напомнить тому, кто будет смотреть отчёт, что тут не тяп-ляп и в продакшн, а была проведена серьёзная работа, глубокий анализ, <s>зарплату аналитику хорошо бы поднять</s>, кратко напомнить о том, что было достигнуто, читатель может быть к концу уже об этом забыл.
</div>


### Выводы:

В ходе выполнения проекта мы познакомились с данными: посмотрели содержимое таблиц и заодно проверили соединение с сервером.

Анализируя содержимое предоставленной БД получили следующую информацию:
- С начала 2000 года вышло 821 книга из присутствующих БД, после 01 января 2000 года вышло 819 книг из присутствующих БД;
- Вывели каждой книги количество обзоров и среднюю оценку, упорядочив по средней оценке;
- Наибольшее число книг толще 50 страниц выпустило издательство Penguin Books, кол-во выпущенных этим издательством книг попадающих под заданный критерий составляет 42;
- Автор с самой высокой средней оценкой книг (учитывая только книги с 50 и более оценками) - J.K. Rowling/Mary GrandPré, среднее средних оценок его книг (с 50 и более оценками) - 4.28;
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.33.

![Апплодисменты](https://i.gifer.com/DhG5.gif "Хлоп-хлоп!")

<div class="alert alert-success"><b>Комментарий ревьюера 2</b>&nbsp;👍<br>
🎉 Поздравляю с успешным завершением последнего проекта финального спринта! 🎂
</div>